CPU Feed-Forward Agent

asmaalrawi opened the pull request:
Adding CPU Feed-Forward Agent, using neural nets to decide frequency online in order to save energy with minimal performance degradation.

alawibaba committed:
CPU Feed Forward Agent

alawibaba committed:
CPU Feed Forward Agent

alawibaba committed:
added regression tester

alawibaba committed:
added regression tester

asmaalrawi committed:
Updated signal names in CPUFFNetAgent for dev rebase

asmaalrawi committed:
Updated signal names in CPUFFNetAgent for dev rebase

lhlawson commented on tutorial/ffnet_agent/CPUFFNetAgent.cpp:
```diff
@@ -0,0 +1,431 @@
+
+        //Save recommendation and convert to SI units for later application
+        package_freq_request.push_back(output[0] * 1e9);
+//        std::cout << "CPUFFNet requested " << output[0] * 1e9 << std::endl;
```
```suggestion
```

lhlawson commented on tutorial/ffnet_agent/CPUFFNetAgent.cpp:
```diff
@@ -0,0 +1,431 @@
+    for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+        //NAN --> Max Frequency
+        if(std::isnan(package_freq_request.at(domain_idx))) {
+            package_freq_request.at(domain_idx) = in_policy[M_POLICY_CPU_FREQ_MAX];
```
If CPU_FREQ_MAX isn't used for anything other than the NAN case is there a reason to provide it in via policy vs relying on the CPU_FREQUENCY_MAX_AVAIL signal?

lhlawson commented on tutorial/ffnet_agent/CPUFFNetAgent.cpp:
```diff
@@ -0,0 +1,431 @@
+
+    // Check for NAN to set default values for policy
+    if (std::isnan(in_policy[M_POLICY_CPU_FREQ_MIN])) {
+        in_policy[M_POLICY_CPU_FREQ_MIN] = min_freq;
```
CPU_FREQ_MIN is unused, may want to remove as a simplification

lhlawson commented on tutorial/ffnet_agent/CPUFFNetAgent.hpp:
```diff
@@ -0,0 +1,135 @@
+        std::map<uint64_t, std::vector<std::pair<std::string, std::string> > > report_region(void) const override;
+        std::vector<std::string> trace_names(void) const override;
+        void trace_values(std::vector<double> &values) override;
+        std::vector<std::function<std::string(double)> > trace_formats(void) const override;
```
There's a missing enforce_policy function that will cause an error when rebased on tip of dev.  You should be able to grab the relevant code from the monitor agent.

Commit:
Fixed double free in LocalNeuralNet destructor.

Commit:
Fixed double free in LocalNeuralNet destructor.

Commit:
Removed max and min frequency from policy.

Commit:
Removed max and min frequency from policy.

Commit:
- Fixed license headers
- Separated out LocalNeuralNet, TensorTwoD, TensorOneD classes
- Fixed array indexing error in LocalNeuralNet

Commit:
- Fixed license headers
- Separated out LocalNeuralNet, TensorTwoD, TensorOneD classes
- Fixed array indexing error in LocalNeuralNet

Commit:
Adjusted agent.py to grab the list of supported agents at runtime.

Signed-off-by: Ali H Mohammad <ali.mohammad@intel.com>
Signed-off-by: Asma H Al-Rawi <asma.h.al-rawi@intel.com>

Commit:
Using existing names function in agents.py for agent discovery.

Commit:
Merge branch 'ahalrawi-cpuffnet' of github.com:asmaalrawi/geopm into ahalrawi-cpuffnet

Commit:
Fixed references to LocalNeuralNet and TensorOneD in CPUFFNetAgent

Commit:
Adjusted input parameters to ffnet integration experiment script:
- Removed unused cpu_fmin and cpu_fmax params
- Added phi_min, phi_max, and phi_step

Signed-off-by: Ali H Mohammad <ali.mohammad@intel.com>

alawibaba committed:
Added unit tests for Local Neural Net library.

alawibaba committed:
More clean.

Commit:
Changed test Makefile to fix coverage compile error.
